cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+                    recv_messages.emplace_back(std::move(msg));
+                }
+            }
+
+            // notify send_tasks to copy & send messages
+            co_await data_ready.notify_all();
+
+            // purge completed send_tasks. This will reset the messages to empty, so that
+            // they release the memory, however the deque is not resized. This guarantees
+            // that the indices are not invalidated. intentionally not locking the mtx
+            // here, because we only need to know a lower-bound on the last completed idx
+            // (ch_next_idx values are monotonically increasing)
+            while (purge_idx + 1 < last_completed_idx) {
+                recv_messages[purge_idx].reset();
+                purge_idx++;
+            }
+            logger.trace("recv_messages active size: ", recv_messages.size() - purge_idx);
+        }
+
+        co_await ch_in->drain(ctx.executor());
+        logger.trace("Process input task completed");
+    }
+
+    coro::mutex mtx;  ///< notify send tasks to copy & send messages
+    coro::condition_variable
+        data_ready;  ///< notify send tasks to copy & send messages notify this task to
+                     ///< receive more data from the input channel
+    coro::condition_variable
+        request_data;  ///< notify recv task to receive more data from the input channel
+    bool input_done{false};  ///< set to true when the input channel is fully consumed
+    std::deque<Message>
+        recv_messages;  ///< messages received from the input channel. We use a deque to
+                        ///< avoid references being invalidated by reallocations.
+    std::vector<size_t> ch_next_idx;  ///< next index to send for each channel
+};
+
+/**
+ * @brief Broadcast messages from one input channel to multiple output channels.
+ *
+ * This is an all-purpose implementation that can support consuming messages by the
+ * channel order or message order. Output channels could be connected to
+ * single/multiple consumer nodes. A consumer node can decide to consume all messages
+ * from a single channel before moving to the next channel, or it can consume messages
+ * from all channels before moving to the next message. When a message has been sent
+ * to all output channels, it is purged from the internal deque.
+ *
+ * @param ctx The context to use.
+ * @param ch_in The input channel to receive messages from.
+ * @param chs_out The output channels to send messages to.
+ * @return A node representing the unbounded fanout operation.
+ */
+Node unbounded_fanout(
+    std::shared_ptr<Context> ctx,
+    std::shared_ptr<Channel> ch_in,
+    std::vector<std::shared_ptr<Channel>> chs_out
+) {
+    auto& executor = *ctx->executor();
+
+    ShutdownAtExit ch_in_shutdown{ch_in};
+    ShutdownAtExit chs_out_shutdown{chs_out};
+    co_await ctx->executor()->schedule();
+    auto& logger = ctx->logger();
+    auto fanout = std::make_unique<UnboundedFanout>(chs_out.size());
```

Comment:
fanout never escapes this function, so can be stack allocated:


```suggestion
    auto fanout = UnboundedFanout(chs_out.size();
```